src/sync/sync_session.cpp

```diff
@@ -283,6 +283,56 @@ SyncSession::SyncSession(SyncClient& client, std::string realm_path, SyncConfig
 , m_realm_path(std::move(realm_path))
 , m_client(client) { }
 
+void SyncSession::handle_progress_update(uint64_t downloaded,
+                                         uint64_t downloadable,
+                                         uint64_t uploaded,
+                                         uint64_t uploadable)
+{
+    std::vector<std::function<void()>> invocations;
+    {
+        std::lock_guard<std::mutex> lock(m_progress_notifier_mutex);
+        m_current_uploadable = uploadable;
+        m_current_downloadable = downloadable;
+        m_current_downloaded = downloaded;
+        m_current_uploaded = uploaded;
+        for (auto& pair : m_notifiers) {
+            bool should_delete = false;
+            invocations.emplace_back(create_notifier_invocation(pair.second, &should_delete));
+            if (should_delete) {
+                m_notifiers.erase(pair.first);
+            }
+        }
+    }
+    // Run the notifiers only after we've released the lock.
+    for (auto& invocation : invocations) {
+        invocation();
+    }
+}
+
+// Create a notifier invocation.
+// Precondition: all four download/upload status member variables have been set to their most up-to-date values.
+std::function<void()> SyncSession::create_notifier_invocation(const NotifierPackage& package,
+                                                              bool* is_expired)
+{
+    REALM_ASSERT(is_expired);
+    uint64_t transferred = (package.direction == NotifierType::download ? m_current_downloaded : m_current_uploaded);
+    uint64_t transferrable = (package.is_streaming
+                              ? (package.direction == NotifierType::download
+                                 ? m_current_downloadable
+                                 : m_current_uploadable)
+                              : package.captured_transferrable);
+    auto invocation = [=](){
+        package.notifier(transferred, transferrable);
+    };
+    if (!package.is_streaming && package.captured_transferrable <= transferred) {
+        // A notifier is expired if the captured max is no longer greater than the current value.
+        *is_expired = true;
+    } else {
+        *is_expired = false;
```

Comment:
I find it slightly easier to parse this by rewriting the condition slightly, to so that we're expired if we've transferred more than was transferrable. You can also just assign the result of the boolean expression directly to `is_expired`:

```c++
is_expired = !package.is_streaming && transferred > package.captured_transferrable;
```